flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala

```diff
@@ -0,0 +1,463 @@
+    /** auxiliary fields for normalized key support */
+    private val auxiliaryFields: (List[Int], Int, Int))
+  extends CompositeTypeComparator[Row] with Serializable {
+
+  // null masks for serialized comparison
+  private val nullMask1 = new Array[Boolean](serializers.length)
+  private val nullMask2 = new Array[Boolean](serializers.length)
+
+  // cache for the deserialized key field objects
+  @transient
+  private lazy val deserializedKeyFields1: Array[Any] = instantiateDeserializationFields()
+
+  @transient
+  private lazy val deserializedKeyFields2: Array[Any] = instantiateDeserializationFields()
+
+  // null checker for reference comparison
+  private val nullChecker = new NullChecker()
+
+  // create auxiliary fields
+  private val normalizedKeyLengths: Array[Int] = auxiliaryFields._1.toArray[Int]
+  private val numLeadingNormalizableKeys: Int = auxiliaryFields._2
+  private val normalizableKeyPrefixLen: Int = auxiliaryFields._3
+
+  // first comparator decides invert the key direction
+  private val invertNormKey: Boolean = comparators(0).invertNormalizedKey()
+
+  def this(
+      keyPositions: Array[Int],
+      comparators: Array[TypeComparator[_]],
+      serializers: Array[TypeSerializer[_]]) = {
+    this(
+      keyPositions,
+      comparators,
+      serializers,
+      createAuxiliaryFields(comparators))
+  }
+
+  private def instantiateDeserializationFields(): Array[Any] = {
+    val newFields = new Array[Any](serializers.length)
+    var i = 0
+    while (i < serializers.length) {
+      newFields(i) = serializers(i).createInstance()
+      i += 1
+    }
+    newFields
+  }
+
+  // --------------------------------------------------------------------------------------------
+  //  Comparator Methods
+  // --------------------------------------------------------------------------------------------
+
+  override def getFlatComparator(flatComparators: util.List[TypeComparator[_]]): Unit =
+    comparators.foreach {
+      case ctc: CompositeTypeComparator[_] => ctc.getFlatComparator(flatComparators)
+      case c@_ => flatComparators.add(c)
+    }
+
+  override def compareToReference(referencedComparator: TypeComparator[Row]): Int = {
+    val other: RowComparator = referencedComparator.asInstanceOf[RowComparator]
+    var i = 0
+    try {
+      while (i < keyPositions.length) {
+        val comparator = comparators(i).asInstanceOf[TypeComparator[Any]]
+        val otherComparator = other.comparators(i).asInstanceOf[TypeComparator[Any]]
+
+        val nullCheck = comparator.equalToReference(nullChecker)
+        val nullCheckOther = otherComparator.equalToReference(nullChecker)
+
+        var cmp = 0
+        // both values are null -> equality
+        if (nullCheck && nullCheckOther) {
+          cmp = 0
+        }
+        // one value is null -> inequality
+        // order is considered for basic types
+        else if (nullCheck || nullCheckOther) {
+          if (comparator.isInstanceOf[BasicTypeComparator[_]]) {
```

Comment:
You can access the sort orders of the individual fields by overwriting the `createComparator` method in the `RowTypeInfo` and directly passing this information to the `RowComparator`.
Then you can avoid the `isInstanceOf` check for basic types and have a generic solution.